File: src/services/CoachingPlanService.js
import dayjs from 'dayjs';
import CoachingPlan from '../model/CoachingPlan';
import {get, put} from './api';

function toPayload(plan) {
    return {
        goal: plan.goal.trim(),
        principles: plan.principles,
        priorities: plan.priorities,
        actions: plan.actions,
        startDate: dayjs(plan.startDate).format('YYYY-MM-DD'),
        reviewDate: plan.reviewDate ? dayjs(plan.reviewDate).format('YYYY-MM-DD') : null,
        notes: plan.notes?.trim() || null
    };
}

export default {
    async get() {
        const data = await get('/coaching-plan');
        return data ? new CoachingPlan(data) : null;
    },
    async save(plan) {
        return new CoachingPlan(await put('/coaching-plan', toPayload(plan)));
    }
};

File: backend/src/main/java/com/jllado/weightcontrol/service/CoachingPlanService.java
package com.jllado.weightcontrol.service;

import com.jllado.weightcontrol.api.dto.CoachingPlanDtos.CoachCoachingPlanRequest;
import com.jllado.weightcontrol.api.dto.CoachingPlanDtos.CoachingPlanRequest;
import com.jllado.weightcontrol.domain.CoachingPlan;
import com.jllado.weightcontrol.domain.User;
import com.jllado.weightcontrol.repository.CoachingPlanRepository;
import jakarta.transaction.Transactional;
import java.util.Optional;
import org.springframework.stereotype.Service;

@Service
@Transactional
public class CoachingPlanService {

    private final CoachingPlanRepository repository;

    public CoachingPlanService(CoachingPlanRepository repository) {
        this.repository = repository;
    }

    public Optional<CoachingPlan> find(User user) {
        return repository.findByUser(user);
    }

    public CoachingPlan replace(User user, CoachingPlanRequest request) {
        validateDates(request);
        CoachingPlan plan = repository.findByUser(user).orElseGet(() -> {
            CoachingPlan created = new CoachingPlan();
            created.setUser(user);
            return created;
        });
        plan.setGoal(request.goal());
        plan.setPrinciples(request.principles());
        plan.setPriorities(request.priorities());
        plan.setActions(request.actions());
        plan.setStartDate(request.startDate());
        plan.setReviewDate(request.reviewDate());
        plan.setNotes(request.notes());
        return repository.save(plan);
    }

    public CoachingPlan replaceConfirmed(User user, CoachCoachingPlanRequest request) {
        if (!request.confirmed()) {
            throw new BadRequestException("Coaching plan write requires explicit confirmation");
        }
        return replace(user, request.plan());
    }

    private void validateDates(CoachingPlanRequest request) {
        if (request.reviewDate() != null && request.startDate().isAfter(request.reviewDate())) {
            throw new BadRequestException("Coaching plan start date must not be after the review date");
        }
    }
}
